Refactor ZooKeeper connection logic out into ZooKeeperLoadBalancer.

incfly opened the pull request:
https://github.com/ClickHouse/ClickHouse/issues/55110

After discussion, this PR is refactor the keeper host selection logic out, into a ZooKeeper Load Balancer. This makes it easier to add features we want (tracking errors from a specific host; find better choice and swap the keeper client when possible; doing all these for availability zone aware load balancing).

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Refactor out the ZooKeeper host selection logic into a separate class `ZooKeeperLoadBalancer`. No-op.

incfly committed:
wip note.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

wip started prototype

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
basic working, use the connect & shuffle & reconnect approach.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
working, no exception handling (fail-over).

az-zoo1, az-zoo2, connected correctly.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
working e2e mostly including fallback.

fallback needs to check the local az is not available case.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
working now not just i != 0 for optimal

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
optimize via static getAvailablityZoneByHost()

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
notes for next optimize, and LoadBalancingOptions:Keeper_AZ

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
working change to a simpler approach.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
clean up

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

clean up comments.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #56563 from incfly/ClickHouse:server-az -->
*This is an automated comment for commit 2efe7fc4bddc9a7455d97d0b113aea52af881249 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56563/2efe7fc4bddc9a7455d97d0b113aea52af881249/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

incfly commented:
@antonio2368 @kssenii  I still need to do lots of polishment, addressing TODO  in the code, adding test, but I have locally tested with 3 keeper & one server, worked as intended. just fyi.

incfly commented on src/Common/ZooKeeper/ZooKeeperImpl.cpp:
```diff
@@ -358,6 +358,8 @@ ZooKeeper::ZooKeeper(
         initFeatureFlags();
         keeper_feature_flags.logFlags(log);
 
+        initAvailabilityZone();
```
Consider to invoke this only local az load balancing option is used.

incfly replied:
Decide not to do. Similar to api flag. separation of concern as well. existing integration test would go through this flow to ensure no issue.

incfly commented on src/Common/ZooKeeper/ZooKeeper.cpp:
```diff
@@ -67,54 +130,64 @@ void ZooKeeper::init(ZooKeeperArgs args_)
-        {
-            auto & host_string = host.host;
-            try
+            // TODO: add DNS check as well.
```
Will restore the DNS logic below and incorporate that to local AZ flow as well..

incfly replied:
Actually not much diff, Github is confused with indentation.

incfly commented on src/Common/ZooKeeper/ZooKeeper.cpp:
```diff
@@ -55,6 +62,62 @@ static void check(Coordination::Error code, const std::string & path)
+    az_by_host[host] = availability_zone;
+}
+
+std::vector<std::string> ZooKeeperAvailabilityZoneMap::shuffleHosts(const std::string & local_az, const std::vector<std::string> & hosts)
```
Return string is not enough. ZooKeeperImpl requires original index in the https://github.com/ClickHouse/ClickHouse/blob/136efba321af48c985b17d97d4108ea00aa56841/tests/integration/test_zookeeper_fallback_session/test.py#L62

incfly replied:
fixed.

incfly committed:
fix original_index missing caused segfault and basic integration test.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

CheSema commented:
Don't you mind if I ask some questions for better mine understanding and further discussions? 

Does the keeper client store the availability flag/information for each server host?
That information is valuable. It helps not to reconnect from not-perfect server when the perfect servers are down.

CheSema commented on src/Common/ZooKeeper/ZooKeeperImpl.cpp:
```diff
@@ -449,14 +452,7 @@ void ZooKeeper::connect(
-                    UInt32 session_lifetime_seconds = fallback_session_lifetime_distribution(thread_local_rng);
-                    client_session_deadline = clock::now() + std::chrono::seconds(session_lifetime_seconds);
-
+                    auto session_lifetime_seconds = setClientSessionDeadline(args.fallback_session_lifetime.min_sec, args.fallback_session_lifetime.max_sec);
```
do I understand that right?
Client make a connection with limitation session life time. After that timeout the client disconnects from that node and tries to choose the better one.

CheSema replied:
What is the value of that `fallback_session_lifetime`?

I see that would be a scenario when one keeper server is down, the clients from the same az are in the loop. They disconnect from less preferred server, trying to connect to the most preferred server, that connection takes time to establish, after that they reconnect to some unpreferred server again. I see it as no positive result, just spend time to trying make some connections.
Do we intent to address such scenario?
I see it like that. Untill client see the better node, client does not unconnect from current one. That approach is challenging. It requires some back ground activity to maintain the liveness of other servers.

CheSema replied:
One more interesting scenario.
When local keeper is down. All clients have to choose different nonlocal servers. I see that this code would choose the one lexical lower nonlocal server.

incfly replied:
> ... to the most preferred server, that connection takes time to establish, after that they reconnect to some unpreferred server again.

That should not happen right?  `setClientSessionDeadline` is only invoked if we are connecting to a server not in the same AZ (less preferred). In case of trying an same-AZ keeper host, 

1. at ZooKeeper.cpp Line 252, the `nodes{node}` is initialized as a single element. So i == 0 condition still applies. We would not call  `setClientSessionDeadline` within `ZooKeeperImpl::connect` 
2. at Line 255, we found local_az == connected_host_az, return directly.

i == 0 is helpful to determine connection is optimal or not in other load balancing scenario as priority can be determined beforehand. In AZ aware load balancing case, I made it `nodes{}` is always inited with one element.

incfly replied:
> All clients have to choose different nonlocal servers.

Good thing is that `ShuffleHost` already addresses this. We call `randomize()` here, https://github.com/ClickHouse/ClickHouse/pull/56563/files#diff-17249c3158c8311e98425540c1d48298b695b41fadb9081cafc203a3a0edd0f9R103

And use ShuffleHost::compare for sorting, which first use the priority and then host.random value for comparision.

https://github.com/ClickHouse/ClickHouse/blob/329e7b5d0a482f4e0c18256b1bc817f9f562019f/src/Common/ZooKeeper/ZooKeeper.h#L64

CheSema replied:
I missed that there is not just host name but a struct `struct ShuffleHost`, thanks for pointing.
I see that it does randomization for equal priorities.

incfly replied:
True. Though that can be mitigated by having a fallback_session_lifetime (say 1 - 3 hours?) that is greater than the time that keeper not being available. Would prefer to keep as it is. It also make things simpler.

CheSema replied:
Do we support cases when there is several servers in one az?
If we do, it is worth to check what connections distribution would be between locals servers in case of its restarts.

incfly replied:
> Do we support cases when there is several servers in one az?
Not yet. Do you mean it could cause overloading the remaining keeper server? that's true. will document that caveat.

CheSema replied:
There is no procedure how the same az connections are balancing. So yeah, I suspect the case when one server holds all the connections without rebalancing.

CheSema commented on src/Common/ZooKeeper/ZooKeeper.cpp:
```diff
@@ -152,6 +225,50 @@ void ZooKeeper::init(ZooKeeperArgs args_)
+            }
+            // Non optimal case that we need to set a session deadline.
+            auto session_timeout_seconds = impl->setClientSessionDeadline(args.fallback_session_lifetime.min_sec, args.fallback_session_lifetime.max_sec);
+            bool need_try_other_host = az_helper.needTryOtherHost(local_az, attempted_hosts);
```
That place is interesting.
As I read this I see that at first run, when client does not know servers zones, the client has to connect to all the servers, because there are some unvisited hosts with unknow az in the list. That could stop early if server with local has been found.
But if we found such local server fast, when it is gone, we will connection to the all rest servers.
Thinks are better when we already know all the servers zones.

CheSema replied:
But I do not understand yet what pice of code is working when client reconnects.
Is `ZooKeeper::init` called for it?

incfly replied:
>  at first run, when client does not know servers zones, the client has to connect to all the servers

Yes initially client do not know server zones. but see line 252, if we are lucky, on the first attempt, client already connects to a local az keeper host, it stops, and not call `setClientSessionDeadline`.


>  if we found such local server fast, when it is gone, we will connection to the all rest servers

Yes. When local az keeper host is gone, we will try other keepers, in a sorted order (local az is first tried, unknown second, and different az last). as long as one connected, keeper client still get initiazlied.

https://github.com/ClickHouse/ClickHouse/pull/56563/files/2efe7fc4bddc9a7455d97d0b113aea52af881249#diff-17249c3158c8311e98425540c1d48298b695b41fadb9081cafc203a3a0edd0f9R94

> But I do not understand yet what pice of code is working when client reconnects.
Is ZooKeeper::init called for it?

Yes. ZooKeeper::startNewSession() -> Constructor -> and then ZooKeeper::init is always invoked. 

https://github.com/ClickHouse/ClickHouse/blob/2b7346d755cf003bf3ba87bac7df275c568080d4/src/Interpreters/Context.cpp#L2920

incfly replied:
> But if we found such local server fast, when it is gone, we will connection to the all rest servers.

Yeah we will try to connect to other hosts whose AZ are unknown yet. But once we know their AZ, `needTryOtherHost` will give different answer. do not bother to retry because we know the rest unattempted hosts are in different AZ anyway.

incfly commented on src/Common/ZooKeeper/ZooKeeper.h:
```diff
@@ -2,6 +2,7 @@
 
 #include "Types.h"
 #include <Poco/Util/LayeredConfiguration.h>
+#include <optional>
```
Note: I want to add a query to expose whether connection is within local az or not. There're two ways to do this:

- Add another field `system.zookeeper_connection` such as `connected_to_local_az`.
- The monitoring system can write a query as below. 

```sql
 :) select 'server-az-eg-us-west-2b' AS f  where f in (select value from system.zookeeper where path = '/keeper' and name ='api_version');

SELECT 'server-us-west-2b' AS f                     
WHERE f IN (
    SELECT value                                                                                                                                                                                                                            
    FROM system.zookeeper
    WHERE (path = '/keeper') AND (name = 'availability_zone')
)           
                                                           
Query id: f4aedc26-259d-49e6-b85f-659a282693ac

```

I am slightly prefer the first approach (add colum for `system.zookeeper_connection`) because that avoids additional request to keeper. WDTY? @antonio2368 @CheSema

incfly replied:
@kgoralski mentioned that we can also use `system.metrics` similar to what he did in https://github.com/ClickHouse/ClickHouse/pull/52378/files

CheSema replied:
I would prefer to find fields `server_availability_zone` and `client_availability_zone` in table `system. zookeeper_connection`.

antonio2368 replied:
I also agree with `system. zookeeper_connection`

incfly replied:
Done. I ended up doing both. Was told it'll be easier to setup using metrics in our case.

incfly committed:
more test cases.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
add more integration test and new config level.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented on src/Common/ZooKeeper/ZooKeeper.cpp:
```diff
@@ -55,6 +62,63 @@ static void check(Coordination::Error code, const std::string & path)
+    return shuffled_hosts;
+}
+
+bool ZooKeeperAvailabilityZoneMap::needTryOtherHost(const std::string &local_az, const std::set<std::string> &attempted_host)
```
NOTE: make sure to document the caveat. if there're multiple keeper hosts, only focus on local az potentially could overload the remaining keeper host(s).

incfly replied:
Done.

incfly committed:
address dns error

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
add documentation and fix original_index.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
metrics and system.zookeeper_connection

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented on src/Common/ZooKeeper/ZooKeeper.cpp:
```diff
@@ -62,6 +64,68 @@ static void check(Coordination::Error code, const std::string & path)
+        LOG_ERROR(log, "Cannot use ZooKeeper host {} due to DNS error: {}", host, e.displayText());
+        return false;
+    }
+    // NOTE: check other exception case how it's handled before.
```
Check this.(to myself)

incfly replied:
https://github.com/ClickHouse/ClickHouse/blob/c465933b1d0173422df6a9fadaec4ba2f5611f17/src/Common/ZooKeeper/ZooKeeper.cpp#L103
same only these two exceptions.

incfly commented on src/Common/ZooKeeper/ZooKeeper.cpp:
```diff
@@ -135,56 +206,17 @@ void ZooKeeper::init(ZooKeeperArgs args_)
-                if (dns_error)
-                    throw KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot resolve any of provided ZooKeeper hosts due to DNS error");
-                else
-                    throw KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot use any of provided ZooKeeper nodes");
```
As you can see, the `Coordination::Error::ZCONNECTIONLOSS` error code here is wrong. I dig a bit, supposed to return bad argument in case of host not found. which is treated as hardware error. but I don't want to introduce diff in this PR.

incfly committed:
cleanup, dns_error_occurred renaming and unittest.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented:
With https://github.com/ClickHouse/ClickHouse/pull/56715 merged,  this should be ready to review now, @CheSema  thanks!

CheSema commented:
Lets add a `warning` message when client sees several keepers from the same az with `keeper_local_availability_zone` enabled.

incfly committed:
fix typo and warn if more than two keeper in same az

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented:
> Lets add a warning message when client sees several keepers from the same az
Done! Could you also help with adding the `ok-test` label?

antonio2368 reviewed (commented):
I have to agree with @CheSema, to make this feature usable we need to have a more complex logic when it comes to switching connections.
First, this doesn't seem like other algorithms because we need to fallback to a different algorithm if we have multiple Keepers in local or non local AZ (it needs to be also configurable).
Second, I agree with @CheSema that we cannot just disconnect on deadline. Local Keeper node could be unavailable for a longer time and I don't like guessing the correct deadlines. Larger deadline would just mean that a node would be "stuck for a longer time on a suboptimal Keeper.
I think we need something more complex like trying to establish a connection to local Keeper, if it succeeds, throw the old connection and replace it with new connection we already made, otherwise keep the connection.
Also, a Keeper in crashloop could pull connection to itself and than just close it.

incfly commented:
> other algorithms because we need to fallback to a different algorithm if we have multiple Keepers in local or non local AZ (it needs to be also configurable).

Okay. While more keeper in a single AZ deployment scenario, in real world can be addressed via infrastructure [feature](https://kubernetes.io/docs/concepts/services-networking/topology-aware-routing/#safeguards), but to make the feature itself complete and robust, I agree we can add that.

I can add that. I am thinking of this: when local AZ Keeper is dropping to a certain threshold, say 80%, then each ClickHouse would take 20% chance to fail over to non local AZ Keepers using random selection. The threshold 80% is configurable. We can have another threshold 

Just want to call out a tricky case: we initialize the host with UNKNOWN, and during ClickHouse startup some Keeper might already be unavailable. I would not kick-in fail-over mechanism for startup. Because each server does not know the full picture. Unless we go with an extreme safety: always initialized all Keeper's AZ information and then start with picking up Keeper to connect; that slows down the startup (every startup doing N^2 connections). This I want to avoid.

> we cannot just disconnect on deadline. would be "stuck for a longer time on a suboptimal Keeper.

Question, why we want to handle this differently for availability zone aware load balancing? The same argument applies to existing load balancers as well. When closest host distance Keeper is not available, we wait for fallback deadline for reconnect. I feel it's an improvement but a general/orthogonal problem.

> Also, a Keeper in crashloop could pull connection to itself and than just close it.

That's a valid point. To address that, we need to let ClickHouse memorize past N connection duration, to detect CrashLoop situation. And temporarily evict that unhealthy Keeper. Or any other ideas?

incfly committed:
Apply suggestions from code review

Co-authored-by: Antonio Andelic <antonio2368@users.noreply.github.com>

incfly commented on src/Common/ZooKeeper/ZooKeeper.cpp:
```diff
@@ -1,15 +1,19 @@
 #include "ZooKeeper.h"
 #include "Coordination/KeeperConstants.h"
 #include "Coordination/KeeperFeatureFlags.h"
+#include "Core/SettingsEnums.h"
```
Done. I will try my best to follow in future!

incfly commented on src/Common/ZooKeeper/ZooKeeper.cpp:
```diff
@@ -152,6 +267,49 @@ void ZooKeeper::init(ZooKeeperArgs args_)
+        attempted_hosts.insert(host.host);
+        try
+        {
+            String host_string = host.host;
```
Removed.

incfly commented on src/Common/ZooKeeper/ZooKeeper.cpp:
```diff
@@ -152,6 +267,49 @@ void ZooKeeper::init(ZooKeeperArgs args_)
+        try
+        {
+            String host_string = host.host;
+            auto host_result = parseForSocketAddress(host_string);
```
Removed. Probably a leftover of previous inline invocation.

incfly commented on src/Common/ZooKeeper/ZooKeeperArgs.cpp:
```diff
@@ -25,6 +26,31 @@ ZooKeeperArgs::ZooKeeperArgs(const Poco::Util::AbstractConfiguration & config, c
+    if (config.hasProperty("availability_zone"))
+    {
+        auto keeper_az = config.getString("availability_zone.value", "");
+        const auto auto_detect_for_cloud = config.getBool("keeper_server.availability_zone.enable_auto_detection_on_cloud", false);
```
Bad copy paste. Fixed.

incfly committed:
fix include quote and others.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
more reference style fix.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented:
@CheSema and I discussed offline, here's the plan:

- We will have a dedicated library(say `KeeperLoadBalancerManager`) manage the state & session. This becomes the source of truth when it comes to pick which Keeper to connect.
- The `KeeperLoadBalancerManager` makes connection to all hosts. Whenever the caller request to connect, it gives the best option at that point. For az-local case, the same AZ keeper would be preferred.
- In background, `KeeperLoadBalancerManager` tracks the score of all keeper host selection score: increase as time passes and still being connected, and reduce the ranking score if there's a disconnection. This should address the CrashLoop case.
- When a better option is available, (e.g. same AZ keeper is back online), this library swap the existing Keeper connection.
- In case of more than one local AZ keeper, we fallback to random selection, but also allow to configure as "sticking with the same AZ regardless". 
- Existing load balancer should be able to migrated to use the same library. We can keep their existing behavior first.

Plan
- I will start with refactoring first, to extract the host selection `i == 0` logic from `ZooKeeperImpl`. This should be no-op.
- In another PR, add basic implementation above for this `KeeperLoadBalancerManager` to support local-az keeper load balancer.

@CheSema let me know if I miss anything. @antonio2368 let me know if you have any suggestions & thoughts.
Thanks!

incfly committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into server-az

incfly committed:
wip basic refactoring.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
shufflehost, test_zookeeper_fallback pass.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
wip nodes -> node and callback

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
Remove AZ code and BetterKeeperHostUpdater as e.g.

ZooKeeperLoadBalancerManager own ZooKeeperImpl: must be shared_ptr.
- reference has to be init during ZooKeeper constructor, infeasible.
Tracking via Callback: more coordination btwn class.. but doable.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
use init not uniq ptr and fix util progm

the declaration is before impl to avoid the situation that
load_balancer_ is destructed first, and then impl trying to record the
callback, it fails. so use field, not uniq ptr.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
renaming and make singleton.

singleton: because ZooKeeper can init multiple times. our data for
tracking however needs to outlive each session.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented:
Hi @CheSema This PR is ready for review now.

CheSema commented on src/Common/ZooKeeper/ZooKeeperLoadBalancer.cpp:
```diff
@@ -0,0 +1,159 @@
+        try
+        {
+            auto client  = std::make_unique<Coordination::ZooKeeper>(host_info_list[i].toZooKeeperNode(), args, zk_log);
+            // Non optimal case: we connected to a keeper host in different availability zone, so set a session deadline.
```
Do you introduce the local az policy in that PR?

incfly replied:
Updated the comment.

CheSema commented on src/Common/ZooKeeper/ZooKeeperLoadBalancer.cpp:
```diff
@@ -0,0 +1,159 @@
+        {
+            auto client  = std::make_unique<Coordination::ZooKeeper>(host_info_list[i].toZooKeeperNode(), args, zk_log);
+            // Non optimal case: we connected to a keeper host in different availability zone, so set a session deadline.
+            if (i != 0)
```
If random policy is active, do we call `setClientSessionDeadline` if just fist random node is out?

incfly replied:
Fixed.

CheSema replied:
What about other policies? For example RoundRobin.

incfly replied:
Good point. RB will iterate from the next (or third) keeper host anyway. Changed to not set connection deadline as well. Others like FIRST_OR_RANDOM IN_ORDER does have a preference on other host. So keep them the same.

incfly committed:
review comments and fix 02882_clickhouse_keeper_client_no_confirmation

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into debug

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
empty commit.

incfly committed:
fix args[i] and host_list_info reset.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
fix message output.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented:
- `test_storage_azure_blob_storage` is a known flake.
- I will fix the binary tidy (some unused code can be removed)
- `test_storage_hdfs` is not relevant.
- AST Fuzzer, this I am not so sure. serverlog shows the stacktrace/code execution is near other places, though, https://pastila.nl/?0033fb43/1da3c715148f2234522ca56d279ba4fd#Rg/KH5gstJdcx1/QE08BEw== 
`/build/src/Functions/formatString.h:127:73: runtime error: addition of unsigned offset to 0x55e7f76d5670 overflowed to 0x55e7f76d566f
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /build/src/Functions/formatString.h:127:73 in 
Received signal -3`

incfly committed:
empty commit

incfly committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into server-az

incfly committed:
roundrobin no reconnect.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into server-az

incfly committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into server-az

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented:
Took a look of `HungCheck` error log,  don't find any correlation with the PR. A few other [failure](https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0b1N0YXJ0T2ZIb3VyKGNoZWNrX3N0YXJ0X3RpbWUpIGFzIGQsCmNvdW50KCksIGdyb3VwVW5pcUFycmF5KHB1bGxfcmVxdWVzdF9udW1iZXIpLCBncm91cFVuaXFBcnJheShjb21taXRfc2hhKSwgZ3JvdXBVbmlxQXJyYXkodGVzdF9uYW1lKQpmcm9tIGNoZWNrcyB3aGVyZSAnMjAyMi0xMi0yNScgPD0gY2hlY2tfc3RhcnRfdGltZSBhbmQgdGVzdF9uYW1lIGlsaWtlICclSHVuZyUnIGFuZCB0ZXN0X3N0YXR1cyBpbiAoJ0ZBSUwnLCAnRkxBS1knLCAnRVJST1InKSBncm91cCBieSBkIG9yZGVyIGJ5IGQgZGVzYw==) recently found here.

same for the integration test. [reports](https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0b1N0YXJ0T2ZIb3VyKGNoZWNrX3N0YXJ0X3RpbWUpIGFzIGQsCmNvdW50KCksIGdyb3VwVW5pcUFycmF5KHB1bGxfcmVxdWVzdF9udW1iZXIpLCBncm91cFVuaXFBcnJheShjb21taXRfc2hhKSwgZ3JvdXBVbmlxQXJyYXkodGVzdF9uYW1lKQpmcm9tIGNoZWNrcyB3aGVyZSAnMjAyMi0xMi0yNScgPD0gY2hlY2tfc3RhcnRfdGltZSBhbmQgdGVzdF9uYW1lIGlsaWtlICcldGVzdF9jYXRib29zdF9ldmFsdWF0ZSUnIGFuZCB0ZXN0X3N0YXR1cyBpbiAoJ0ZBSUwnLCAnRkxBS1knLCAnRVJST1InKSBncm91cCBieSBkIG9yZGVyIGJ5IGQgZGVzYw==).

CheSema commented on src/Common/ZooKeeper/ZooKeeperLoadBalancer.cpp:
```diff
@@ -0,0 +1,164 @@
+            // Except for RANDOM and ROUND_ROBIN. For random, as it's random anyway regardless of first or second host.
+            // For round robin, hosts tried later will be in the same order as well.
+            const auto & lb_algorithm = args.get_priority_load_balancing.load_balancing;
+            if (i != 0 && (lb_algorithm != LoadBalancing::RANDOM && lb_algorithm != LoadBalancing::ROUND_ROBIN))
```
I do not like that still.
I think that we should think about desire model here and what interface will be good to implement that model.

I want to see the model when lb_algorithm defines which connections are permanent and which are temporary. For example:  HOSTNAME_LEVENSHTEIN_DISTANCE could have several host with equal priority. We do not what do a loop switching between such equal hosts. That decision should not  be done by `createClient` based on just host order.

CheSema replied:
also ROUND_ROBIN,FIRST_OR_RANDOM have mutable priorities , each call `getPriorityFunc` returns different result.

incfly replied:
Fair point. `HOSTNAME_LEVENSHTEIN_DISTANCE` indeed can have several host with same priority. `i != 0` does not mean we end up with a sub-optimal choice. We need a better interface other than this naive check. Though as discussed before, static algorithms (existing ones) would select hosts purely by algorithm itself; availability zone would require input and interact with ZooKeeperLoadBalancer tracking history. Let me think about that.

Also `getPriorityFunc` is used in clickhouse client connection pool. We probably don't want to change that, for now.

> do we somewhere take getPriorityFunc into account?
Not sure what you meant by that. getPriorityFunc is called outside the loop once per connection, as before.

> also ROUND_ROBIN,FIRST_OR_RANDOM have mutable priorities
ROUND_ROBIN does. But FIRST_OR_RANDOM should be deterministic, we always pass in offset = 0 in the getPriorityFunc call here.
 https://github.com/ClickHouse/ClickHouse/blob/6b35213dd9154dbf773eb1dab0d33e17b8c5df36/src/Common/GetPriorityForLoadBalancing.cpp#L33

CheSema replied:
just get rid off if `getPriorityFunc`.

CheSema replied:
check this @CheSema
[add IClientsConnectionBalancer](https://github.com/ClickHouse/ClickHouse/pull/56563/commits/c189c48b47ed81a13dc9976603db9f802be6457f)

CheSema replied:
@antonio2368 
Since I starting contributing here, I would ask you, please, to review us both. When I open this PR as ready.

CheSema replied:
@incfly please make in draft.

CheSema committed:
add IClientsConnectionBalancer

incfly committed:
ZooKeeperLB::init() only for new config.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented on src/Common/ZooKeeper/ZooKeeperLoadBalancer.cpp:
```diff
@@ -0,0 +1,801 @@
+
+    connection_balancer = getConnectionBalancer(args.get_priority_load_balancing.load_balancing);
+
+    for (size_t i = 0; i < args_.hosts.size(); ++i)
```
@CheSema  In this commit https://github.com/ClickHouse/ClickHouse/pull/56563/commits/ad95bdbf819a96acbe9533299fbfe7e4ea0fd042

I changed it so that 
- `ZooKeeper::startSession` would not re-init the load balancer.
- however, when config reloaded, should be due to `system reload config` after checking code, we will reset;

Additionally as for the question on "do we have concurrent createClient"? I think it could be. We have `auxilary_zookeeper` and can multiple named ZK configuration. https://github.com/ClickHouse/ClickHouse/pull/14155/files#diff-50fdf83595b40f8695c7f6ecdaff323750211e95c8e778f0c8c627912c66115cR17

I think it's better to separate their states with one load balancer instance per config_name. Because while zk host address can be the same, people can configure different timeout/error simulation settings.

After that, mutex within `ZooKeeperLoadBalancer` can be removed.

incfly committed:
fix the zk args, and offline online.

copy paste bugs in both cases.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
fix the round robin case.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
wip no by adress.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
simplify no getPriorities methods.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
fix optimal priorities

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
constpriority class only

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
roundrobin with status inheritance directly

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
nit reportAllEndpoints

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
nit dns error skip; betterhost would work for az.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
resetall status. one leveh distance failed.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
pass two integ test.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
nitpick addEndpoint back.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
remove endpointWorthChecking, might just check all in future.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
additional reload param for build.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
apply logger in lb.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
Merge remote-tracking branch 'origin/master' into git-module

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
use right zk_log and not std move.

std::move after that is nullptr. not sure what i did before.
not caused by setLogger. maybe better to use a log provider func to
avoid maintain too many copies in different classes.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
check only impementation zookeeper

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented:
PR is ready for review, CI tests are also mostly green. @antonio2368 @CheSema could you take a look? Thanks!

CheSema commented:
i have looked this. It looks promising. You have an interface to work with. Your implementation is shorter and more simple than in my example but still powerful and flexible. I like it.

I would like to see some units tests here. They is going to fix the behaviour and provide some example how to work with new interface. I see some usage of a new interface in `src/Common/ZooKeeper/examples/zkutil_test_commands_new_lib.cpp`. But I do not understand. Is it a test or just tool for manual testing?

incfly commented:
Thanks! I touched `src/Common/ZooKeeper/examples/zkutil_test_commands_new_lib.cpp` here because it uses the `ZooKeeperImpl` constructor directly, which we changed in this PR.

I will look at adding some unit test for the library.

incfly committed:
unit test.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into server-az

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented:
Hi @CheSema unit test added. Can you see if anything else is missing?

incfly committed:
fully changed to dependency ingestion.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
template T for swap TestHelper impl.

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into server-az

incfly committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into server-az

incfly committed:
host init

Signed-off-by: Jianfei Hu <hujianfei258@gmail.com>

incfly commented:
Updated with recent master. Friendly ping, @CheSema could you take a look? Thanks!

tavplubix commented on src/Common/ZooKeeper/ZooKeeperLoadBalancer.cpp:
```diff
@@ -0,0 +1,622 @@
+#include <Poco/String.h>
+
+
+enum Status
+{
+    UNDEF = 0,
+    ONLINE,
+    OFFLINE,
+};
```
Do we really need 3 statuses? Once we get a status for a node, it immediately becomes outdated: an online node may suddenly go down, and an offline node may go up quickly. So it feels like the only status is UNDEF, we never know for sure. 

There are 2 main usecases:
 - We are connected to a suboptimal node. We can just try connecting to the optimal one, and switch if the attempt was successful; keep the suboptimal connection otherwise
 - Connection is lost and we need a new one. We can just try connecting to another node because only 1 of 3 nodes can be unavailable at the same time 

In both cases we can just connect optimistically and don't care about cached statuses
